Fix the following typos:

elianerpereira opened the pull request:
import - imported
semver - server
provides - provider
provide - provide
Remove `describes`
this - these
initalization - initialization
Simpliflied - Simplified

stefwalter commented:
This was a result of my "Day of Learning", and had so much fun working with Image Builder. Hope the documentation helps others.

gicmo commented:
This is very cool!!

For the iso creation part, there is actually a tool called `mkksiso` (written by @bcl) that can take an iso and a kickstart to it (it modifies the kernel command line as well). Haven't tried it myself though.

dvdhrm commented on howto/howto-ostree-anaconda.md:
```diff
@@ -0,0 +1,221 @@
+permalink: /howto/howto-ostree-anaconda.md
+---
+
+{%- include ../global.html -%}
```
Is there a reason for that include? I am also unsure how this is supposed to work. Includes in jekyll are relative to `_include/`, so the path `../global.html` looks wrong to me.

stefwalter replied:
No idea. I just copied it from another directory, and added the dots since this is in a subdirectory. I'll try jekyll serve.

dvdhrm replied:
It contains shared objects we use to avoid regenerating lists in jekyll everywhere. You can just omit it here. It can be included when it is needed.

stefwalter replied:
No idea. But I guess I'll try jekyll serve.

dvdhrm commented on howto/howto-ostree-anaconda.md:
```diff
@@ -0,0 +1,221 @@
+
+{%- include ../global.html -%}
+
+# How to: Image Bulider + OSTree + Anaconda
```
`Bulider` -> `Builder`

stefwalter replied:
Fixed.

dvdhrm commented on howto/howto-ostree-anaconda.md:
```diff
@@ -0,0 +1,221 @@
+
+# Prerequisites
+
+The naming of *Image Builder* and it's various parts need some introduction.
```
You say `the naming [...] need[s] some introduction`, but then you don't introduce it? Is that some leftover, or am I just reading it wrong? At least, this confused me, given that you follow it with `Here's how you get it all installed.`.

stefwalter replied:
Whoops. Added some clarity.

dvdhrm commented on howto/howto-ostree-anaconda.md:
````diff
@@ -0,0 +1,221 @@
+The naming of *Image Builder* and it's various parts need some introduction.
+Here's how you get it all installed. We'll also use ```podman``` later on.
+
+    $ sudo yum install osbuild osbuild-ostree osbuild-composer composer-cli podman
````
Can we use `dnf` instead?

stefwalter replied:
DNF is yum 4 ... and in particular that name is not available nor documented on RHEL. My goal is that this documentation should work with Fedora and RHEL.

DNF really ended up being just an interim (ie: "Did Not Finish") working name until YUM 4 was complete. Much like the naming confusion of Image Builder, Composer, OSBuild, Lorax, Weldr ... we're really bad at naming.

dvdhrm commented on howto/howto-ostree-anaconda.md:
```diff
@@ -0,0 +1,221 @@
+You can create OSTree commits nicely in the Image Builder GUI. If
+you'd rather do this via the command line, then skip this section.
+
+    $ sudo yum install cockpit cockpit-composer
```
Same `yum vs dnf` here.

stefwalter replied:
Ditto.

dvdhrm commented on howto/howto-ostree-anaconda.md:
````diff
@@ -0,0 +1,221 @@
+![Image Builder empty screen](howto-ostree-anaconda-empty.png)
+
+Click the *Create Blueprint* button to make a new blueprint and give it
+a simple name. In the instructions below we use ```iot``` so for simplicity
````
You use triple-backticks quite a lot. Is there a reason for that? Single-backticks should work just fine, but that is probably just personal style.

dvdhrm commented on howto/howto-ostree-anaconda.md:
```diff
@@ -0,0 +1,221 @@
+![Image Builder Create Blueprint](howto-ostree-anaconda-create-blueprint.png)
+
+You don't need to add any additional software to the blueprint to get a bootable
+system. Click the *Create Image* button, and choose the *Fedora IoT Commit* as
```
`and choose the` -> `and choose`

dvdhrm commented on howto/howto-ostree-anaconda.md:
```diff
@@ -0,0 +1,221 @@
+
+You don't need to add any additional software to the blueprint to get a bootable
+system. Click the *Create Image* button, and choose the *Fedora IoT Commit* as
+the image *Type*:
```
`the image` -> `image` (but both work here, I think?)

dvdhrm commented on howto/howto-ostree-anaconda.md:
```diff
@@ -0,0 +1,221 @@
+
+Once you click *Create* the image build will start, and take several minutes
+to complete. Unfortunately you won't see any progress information or logs
+until its done, so just be patient:
```
This is intentional, not unfortunate. I would drop that sentence, as it judges the process differently than our internal discussions.

dvdhrm commented on howto/howto-ostree-anaconda.md:
````diff
@@ -0,0 +1,221 @@
+# Setting up the Kickstart
+
+We'll install the OSTree in the Anaconda Installer using a Kickstart file.
+Create a file called ```ostree.ks``` containing the following content:
+in the current directory, the same containing the following content:
````
This needs some fixup.

dvdhrm commented on howto/howto-ostree-anaconda.md:
````diff
@@ -0,0 +1,221 @@
+I've chosen the name ```iot```, which is used in the steps below. For your
+first go around just keep this simple name so you don't have to figure out all
+the places below to change the name. Even though we haven't added any specific
+```packages``` to this blueprint, all the necessary packages to get a
````
GitHub uses lines starting with three backticks as block-quotes and will highlight it as such. I don't think standard jekyll-markdown does it, but maybe we can use single-backticks here to avoid the syntax-highlighting confusion?

dvdhrm commented on howto/howto-ostree-anaconda.md:
```diff
@@ -0,0 +1,221 @@
+OSTree unfortunately doesn't come with a stock way to push or
+serve OSTree commits, so we'll need to build our own mechanism to serve the
+OSTree content. Lets use [Apache httpd](https://httpd.apache.org/) in
+a container. The OSTree client is so agressive with it's HTTP usage
```
`The OSTree client is so agressive with it's HTTP usage and overwhelms many HTTP servers.`

Can you explain what this means?

stefwalter replied:
It opens up so many HTTP connections simultaneously that it overwhelms HTTP servers. Not sure how many, but if I had to guess it would be > 64.

In particular, if you serve this content with ```python -m http.server 8000``` instead of the podman containerthen Anaconda will fail halfway through install due to the HTTP server refusing connections.

On #ostree IRC we had a discussion, and a regular developer of ostree remarked that:

```
<w4tsn[m]> I also experienced that the default nginx config may cause troubles with timeouts on tcp connections so I had to increase certain timeouts / limits. Don't remember which exactly. If someone is interested I could look them up
```

@cgwalters is aware of this point and remarked:

```
<walters> stefw: hi!  yep, w4tsn[m] is right, just use any webserver
<stefw> unfortunately http.server from python3 doesn't work
<stefw> but nodejs http-server does
<stefw> ostree client causes way too many parallel connections for 'python3 -m http.server 8000'
<stefw> Eventually we see an error saying that the connection was rejected
<stefw> But 'npm install http-server' worked fine.
<walters> ohh yeah that
<walters> part of the idea was that http/2 would be deployed earlier than it was (and it definitely works much better with lots of parallel connections) but the other thing that happened is http 2 became "TLS only" for the most part, which then ups the difficulty for "let me just test this locally" case
<walters> 16:13:33> I end up using `kola http-server` mostly which is a thin wrapper in our devtools around the golang webserver stack, but yeah at some point may need to dig into why the python webserver breaks on this
<walters> i guess since python isn't multithreaded we could detect the remove webserver and just not do anything parallel
```

@dvdhrm Would you like me to add further explanation in the document, or is noting this here enough?

dvdhrm replied:
> Would you like me to add further explanation in the document, or is noting this here enough?

Right, I see the reasoning now. Thanks! I think your use of `and` confused me. How about writing `that it overwhelms` rather than `and it overwhelms`?

cgwalters replied:
Can you replace most of this text with a link to https://github.com/ostreedev/ostree/issues/2120 ?

cgwalters replied:
> OSTree unfortunately doesn't come with a stock way to push or serve OSTree commits

I am *strongly* of the opinion the ostree project shouldn't come with a webserver; there's just no reason for that.  I wouldn't call it unfortunate.

Pushing commits around is a huge interesting topic that's also related to the "static content".

But remember, part of the goal of OSTree is to be an *infrastructure* piece or a shared library, not an end-to-end toolkit.  Because we want to support multiple distributions mainly.

But there are plenty of extant projects which both push and serve commits.    coreos-assembler for example *is* intending to be more of an "end-to-end toolkit", and for example we have a complex chain to sign commits integrating with Fedora: https://github.com/coreos/coreos-assembler/blob/master/src/cmd-sign

stefwalter replied:
Thanks for the insight. Good to know about ostree's goals. I've taken out the word "unfortunately".

That said, this why I had to write this document. The individual tools haven't claimed ownership of a solution end to end yet for using rpm-ostree as a general purpose Linux.

I imagine Image Builder is in a position to pull that off. Fingers crossed...

dvdhrm commented on howto/howto-ostree-anaconda.md:
```diff
@@ -0,0 +1,221 @@
+virtual machine or hardware you can do this in different ways that I won't
+cover here.
+
+At the initial screen of that the Installer ISO displays, you'll see a line
```
`that` -> `what`?

stefwalter replied:
Thanks. Fixed.

dvdhrm commented on howto/howto-ostree-anaconda.md:
```diff
@@ -0,0 +1,221 @@
+
+![Anaconda boot menu](howto-ostree-anaconda-bootmenu.png)
+
+Before the time out expires, use the arrow keys to select that line, and press
```
Can a "time-out expire"? This sounds done twice (is there a translation of "doppelt gemoppelt"? There should be!) Time can run out, or a timer expires. Or I am just wrong ;) Regardless, this is overly pedantic, feel free to leave as it is.

stefwalter replied:
Heh. No worries. Removed the extra word wort.

dvdhrm commented on howto/howto-ostree-anaconda.md:
```diff
@@ -0,0 +1,221 @@
+
+![Anaconda kernel command line](howto-ostree-anaconda-kernel-commandline.png)
+
+After you press ENTER, the installer should start. If things are wortking the
```
`wortking` -> `working`

stefwalter replied:
Fixed.

dvdhrm commented on documentation.md:
```diff
@@ -24,6 +24,9 @@ respective user interface is the best source of help. Furthermore, the
 internals of the image building tools.
 
+* **How Tos**:
+   * [**Image Builder + OSTree + Anaconda**](howto/howto-ostree-anaconda.md)
```
This link does not work, does it? This would be relative to `documentation/`. I think you need something like:

```
[**Image Builder + OSTree + Anaconda**]({% link howto/howto-ostree-anaconda.md %})
```

dvdhrm commented:
I reverted the merge, since it breaks website generation. GitHub does not let me re-open the PR, so we probably need to open a new PR to get this fixed and merged again. Sorry for that!

stefwalter commented:
> Thanks a lot! This looks quite nice. A few nitpicks inline. You can also test this locally with `jekyll serve`. The code right now does not compile, and the link to the howto does not resolve.
> 
> I also wonder whether we want this as a blog-entry instead? This would allow us to mark it as guest-entry, add proper attribution and have it included in the feed. We can then still link to the entry from `documentation.md`. What do you think?

Thanks for the review!

That sounds like a good idea. I'm not sure how to do it. But I'll take the first step of opening up a new pull request.

dvdhrm commented:
> That sounds like a good idea. I'm not sure how to do it. But I'll take the first step of opening up a new pull request.

Thanks! I can convert it to a news-entry. I think this would also give you more freedom of putting your own opinion into it. A guest entry can judge things from an outsider view, while I would prefer internal documentation to be aligned with the consensus we have in the project (which can also change over time, obviously).

I might also add a jekyll-build to the CI. I cannot replicate what github-pages does, but I think we can come close enough.

stefwalter commented:
Fixed up in #21

elpereir committed:
Fix the following typos:

import - imported
semver - server
provides - provider
provide - provide
Remove `describes`
this - these
initalization - initialization
Simpliflied - Simplified

ochosi commented on docs/user-guide/01-blueprint-reference.md:
```diff
@@ -13,9 +13,9 @@ version = "0.0.1"
+- The `name` attribute is a string that contains the name of the blueprint. It can contain spaces, but they will be converted to `-` when it is imported into `osbuild-composer`. It should be short and descriptive.
 - The `description` attribute is a string that can be a longer description of the blueprint and is only used for display purposes.
-- The `version` attribute is a string that contains a semver compatible version number. If a new blueprint is uploaded with the same version the server will automatically bump the PATCH level of the version. If the version doesn't match it will be used as is. For example, uploading a blueprint with version set to 0.1.0 when the existing blueprint version is 0.0.1 will result in the new blueprint  ...
+- The `version` attribute is a string that contains a server compatible version number. If a new blueprint is uploaded with the same version the server will automatically bump the PATCH level of the version. If the version doesn't match it will be used as is. For example, uploading a blueprint with version set to 0.1.0 when the existing blueprint version is 0.0.1 will result in the new blueprint  ...
```
It really should be "semver" as in "semantic versioning", not server. However, let's not use the abbreviation but instead "semantic versioning" for more clarity.

elianerpereira replied:
Thank you for clarifying this. I will fix it.

ochosi commented on docs/user-guide/01-blueprint-reference.md:
```diff
@@ -64,7 +64,7 @@ The `packages` and `modules` lists contain objects with a `name` and optional `v
 *Currently there are no differences between packages and modules in `osbuild-composer`. Both are treated like an rpm package dependency.*
 
-> When using virtual provides as the package name the version glob should be `*`. And be aware that you will be unable to `freeze` the blueprint. This is because the provide will expand into multiple packages with their own names and versions.
+> When using virtual providers as the package name the version glob should be `*`. And be aware that you will be unable to `freeze` the blueprint. This is because the provider will expand into multiple packages with their own names and versions.
```
I think this should actually be "provides". In virtual packages, this keyword is used to list all the packages provided by a virtual package.
Maybe we could make it clearer by using markdown for monospace formatting `provides` to make it visually stand out as a technical term?

elianerpereira replied:
Thank you for clarifying this. I will fix it and add the correct markup.

ochosi reviewed (changes requested):
This is awesome - thanks for catching so many typos!
I think there are two false positives there, but either also warrants improvements for more clarity.

elpereir committed:
Fix false positives

server - semantic versioning
providers - `provides`
provider - `provides`

ochosi commented on docs/user-guide/01-blueprint-reference.md:
```diff
@@ -13,9 +13,9 @@ version = "0.0.1"
+- The `name` attribute is a string that contains the name of the blueprint. It can contain spaces, but they will be converted to `-` when it is imported into `osbuild-composer`. It should be short and descriptive.
 - The `description` attribute is a string that can be a longer description of the blueprint and is only used for display purposes.
-- The `version` attribute is a string that contains a semver compatible version number. If a new blueprint is uploaded with the same version the server will automatically bump the PATCH level of the version. If the version doesn't match it will be used as is. For example, uploading a blueprint with version set to 0.1.0 when the existing blueprint version is 0.0.1 will result in the new blueprint  ...
+- The `version` attribute is a string that contains a server compatible version number. If a new blueprint is uploaded with the same version the semantic versioning will automatically bump the PATCH level of the version. If the version doesn't match it will be used as is. For example, uploading a blueprint with version set to 0.1.0 when the existing blueprint version is 0.0.1 will result in the n ...
```
Sorry, but this is now doubly wrong 😅

elianerpereira replied:
Sorry, fixed, hopefully :D

ochosi reviewed (changes requested):
Thanks for fixing the one, one left to go

elpereir committed:
Fix doubly wrong fix

server - semantic versioning
semantic versioning - server

ochosi reviewed (approved):
Great, thanks for all the fixes!
